File: src/solutions/s0001_two_sum.rs
// Time complexity: O(n)
// Space complexity: O(n)
use std::collections::HashMap;

pub struct Solution;

impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let mut num_to_index = HashMap::new();

        for (i, &num) in nums.iter().enumerate() {
            match num_to_index.get(&(target - num)) {
                Some(&index) => return vec![index as i32, i as i32],
                None => num_to_index.insert(num, i as i32),
            };
        }
        panic!();
    }
}

File: src/solutions/s0001_two_sum.cpp
#include <vector>
#include <unordered_map>
#include <stdexcept>

class Solution {
 public:
  std::vector<int> twoSum(const std::vector<int>& nums, int target) {
    std::unordered_map<int, int> numToIndex;

    for (int i = 0; i < nums.size(); ++i) {
      auto it = numToIndex.find(target - nums[i]);
      if (it != numToIndex.end()) return {it->second, i};
      numToIndex[nums[i]] = i;
    }
    throw std::invalid_argument("No two sum solution");
  }
};
